### PR TITLE
slt: Use scratch directory for lgalloc

### PR DESCRIPTION
**Pull request by def-**
Currently SLT is seeing many warnings:
```
WARN mz_compute::compute_state: Not enabling lgalloc, scratch directory not specified
```
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by nrainer-materialize — APPROVED** [review]

**Comment by def-** [comment]
@antiguru I'm now seeing `Mmap pool exhausted, falling back to heap: I/O error` in the logs: https://buildkite.com/materialize/tests/builds/69371#018bf793-15f0-4c3a-9cb8-28e024c55744 Do you have any idea?

**Comment by def-** [comment]
I think the previous logic of creating a tempdir in SLT was wrong since it would get deleted after the config object has been created. I'll try with a fixed `./scratch/` directory for now. Does it have to be a different directory for each server?

Now I got a coredump locally trying to fix this up, starting to get interesting maybe?
```
[580725.853745] timely:work-0[2329748]: segfault at 7f35c0bfb198 ip 0000562e335c1e7e sp 00007f35c0bfb180 error 6 in clusterd[562e2cf32000+b2e1000] likely on CPU 1 (core 1, socket 0)
[580725.853756] Code: 8d 05 b6 80 c2 f7 48 89 c7 b8 00 00 00 00 e8 ff c8 1b 00 b8 00 00 00 00 48 8b 5d f8 c9 c3 55 48 89 e5 53 48 81 ec f8 7d 00 00 <48> 89 bd 18 82 ff ff 48 89 b5 10 82 ff ff b8 00 00 00 00 84 c0 0f
```

File: src/controller/src/clusters.rs [hunk 1]
```diff
@@ -618,6 +619,13 @@ where
                             ));
                         }
 
+                        if let Some(scratch_dir) = &scratch_dir {
+                            args.push(format!(
+                                "--scratch-directory={}",
+                                scratch_dir.to_str().unwrap()
+                            ));
+                        }
+
```
Comment (antiguru) [inline src/controller/src/clusters.rs:628]: I don't think this is correct. We set the `--scratch-directory` in the orchestrator if `disk` is set.

File: src/environmentd/src/bin/environmentd/main.rs [hunk 2]
```diff
@@ -907,6 +907,7 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
             ),
             secrets_reader_aws_prefix: Some(aws_secrets_controller_prefix(&args.environment_id)),
         },
+        scratch_dir: None,
```
Comment (antiguru) [inline src/environmentd/src/bin/environmentd/main.rs:910]: I think you're looking for `orchestrator_process_scratch_directory` instead.

File: src/controller/src/clusters.rs [hunk 3]
```diff
@@ -588,6 +588,7 @@ where
         let persist_pubsub_url = self.persist_pubsub_url.clone();
         let enable_persist_txn_tables = self.enable_persist_txn_tables;
         let secrets_args = self.secrets_args.to_flags();
+        let scratch_dir = self.scratch_dir.clone();
```
Comment (antiguru) [inline src/controller/src/clusters.rs:591]: I think you can get entirely get rid of `scratch_dir`.

**Comment by def-** [comment]
@antiguru Is there a way to make every cluster have disk by default? If disk is required for lgalloc, this means we are not testing it in most of CI currently, right?

**Comment by antiguru** [comment]
There is a feature flag that changes the default value for DISK. However, it cannot distinguish between compute and storage clusters, so it might have undesirable side effects. I tried enabling it, and observed failures in cloud test because the disk requirement could not be fulfilled. The compute and storage teams need to converge on whether all clusters should have disk, essentially turning the DISK keyword into a no-op.

**Comment by def-** [comment]
If it's not easy to currently enable in CI, then I guess we have to live with less coverage of jemalloc. My test runs didn't find anything strange:
https://buildkite.com/materialize/tests/builds/69576#_
https://buildkite.com/materialize/sql-logic-tests/builds/5750#_

**Pull request closed without merging**